kernel/kernel-default/src/main/java/io/delta/kernel/parquet/MapConverter.java

```diff
@@ -169,24 +169,24 @@ public Converter getConverter(int fieldIndex)
         }
 
         @Override
-        public void start()
-        {
-            Arrays.stream(converters)
-                .filter(conv -> !conv.isPrimitive())
-                .forEach(conv -> ((GroupConverter) conv).start());
+        public void start() {
+            for (int i = 0; i < converters.length; i++) {
+                if (!converters[i].isPrimitive()) {
+                    ((GroupConverter) converters[i]).start();
+                }
+            }
         }
 
         @Override
-        public void end()
-        {
-            Arrays.stream(converters)
-                .filter(conv -> !conv.isPrimitive())
-                .forEach(conv -> ((GroupConverter) conv).end());
-
-            Arrays.stream(converters)
-                .map(converter -> (ParquetConverters.BaseConverter) converter)
-                .forEach(converter -> converter.moveToNextRow());
-
+        public void end() {
+            for (int i = 0; i < converters.length; i++) {
+                if (!converters[i].isPrimitive()) {
+                    ((GroupConverter) converters[i]).end();
+                }
+            }
+            for (int i = 0; i < converters.length; i++) {
+                ((ParquetConverters.BaseConverter) converters[i]).moveToNextRow();
```

Comment:
@vkorukanti a few questions
1. why does this `end` method also call `moveToNextRow` ?
2. is there a temporal relationship between the first for loop and second? i.e. must _all_ the converters be `end`ed before any of them call `moveToNextRow`?

Reply:
This is because we are returning a columnar data from a row based reader. The `end` means Parquet is saying row reading is done.

Here are the sequence of events when reading a row:
* Parquet reader calls `start` on all column converters (one for each column being read). 
* Parquet reader makes calls on the column converters to give the value for each column in the row
* Parquet reader calls `end` on all column converters to signal that a row is done reading.

We extend the `ColumnConverter`s with an additional methods `moveToNextRow` which signals the column converters to clean up the current row and get ready for reading the next row (increment the row counter etc.)

Commented [here](https://github.com/delta-io/delta/blob/master/kernel/kernel-default/src/main/java/io/delta/kernel/parquet/ParquetBatchReader.java#L174).